// ang-jsoneditor/esm5/lib.module.js
import { NgModule, CUSTOM_ELEMENTS_SCHEMA, NO_ERRORS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { JsonEditorComponent } from './jsoneditor/jsoneditor.component';
import { FormsModule } from '@angular/forms';
var NgJsonEditorModule = /** @class */ (function () {
    function NgJsonEditorModule() {
    }
    NgJsonEditorModule.forRoot = function () {
        return {
            ngModule: NgJsonEditorModule,
            providers: []
        };
    };
    NgJsonEditorModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule
                    ],
                    declarations: [
                        JsonEditorComponent
                    ],
                    exports: [
                        JsonEditorComponent
                    ],
                    schemas: [
                        CUSTOM_ELEMENTS_SCHEMA,
                        NO_ERRORS_SCHEMA
                    ]
                },] }
    ];
    return NgJsonEditorModule;
}());
export { NgJsonEditorModule };
